File: apps/java/design-pattern/src/main/java/pattern/builder/ComputerBuilder.java
package pattern.builder;

import java.util.ArrayList;
import java.util.List;

/**
 * コンピュータビルダー基底クラス。
 * バリデーション: メモリ 250 以上、ドライブ 1-4 台、ハードディスク必須。
 */
public abstract class ComputerBuilder {
    protected boolean turbo = false;
    protected int memorySize = 512;
    protected final List<Drive> drives = new ArrayList<>();
    protected String display;

    public void setTurbo(boolean turbo) {
        this.turbo = turbo;
    }

    public void setMemorySize(int memorySize) {
        this.memorySize = memorySize;
    }

    public void addCd(boolean writable) {
        drives.add(new Drive("cd", 760, writable));
    }

    public void addDvd(boolean writable) {
        drives.add(new Drive("dvd", 4700, writable));
    }

    public void addHardDisk(int size) {
        drives.add(new Drive("hard_disk", size, true));
    }

    public Computer getComputer() {
        validate();
        String cpu = turbo ? "TurboCPU" : "BasicCPU";
        Motherboard motherboard = new Motherboard(cpu, memorySize);
        return new Computer(display, motherboard, List.copyOf(drives));
    }

    private void validate() {
        if (memorySize < 250) {
            throw new IllegalStateException("Not enough memory: " + memorySize);
        }
        if (drives.size() > 4) {
            throw new IllegalStateException("Too many drives: " + drives.size());
        }
        boolean hasHardDisk = drives.stream()
                .anyMatch(d -> "hard_disk".equals(d.type()));
        if (!hasHardDisk) {
            throw new IllegalStateException("Must have at least one hard disk");
        }
    }
}

File: apps/ruby/design-pattern/15-builder/lib/computer_builder.rb
require_relative "parts"

class ComputerBuilder
  def initialize(display:)
    @display = display
    @turbo = false
    @memory_size = 512
    @drives = []
  end

  def turbo(has_turbo = true)
    @turbo = has_turbo
  end

  def memory_size=(size)
    @memory_size = size
  end

  def add_cd(writable = false)
    @drives << Drive.new(:cd, 760, writable)
  end

  def add_dvd(writable = false)
    @drives << Drive.new(:dvd, 4_700, writable)
  end

  def add_hard_disk(size)
    @drives << Drive.new(:hard_disk, size, true)
  end

  def computer
    validate!
    cpu = @turbo ? TurboCPU.new : BasicCPU.new
    motherboard = Motherboard.new(cpu, @memory_size)
    Computer.new(@display, motherboard, @drives)
  end

  private

  def validate!
    raise "Not enough memory: #{@memory_size}" if @memory_size < 250
    raise "Too many drives: #{@drives.size}" if @drives.size > 4
    raise "Must have at least one hard disk" unless @drives.any? { |drive| drive.type == :hard_disk }
  end
end

class DesktopBuilder < ComputerBuilder
  def initialize
    super(display: :crt)
  end
end

class LaptopBuilder < ComputerBuilder
  def initialize
    super(display: :lcd)
  end

  def computer
    raise "Laptop display must be LCD" unless @display == :lcd
    super
  end
end
